13_07_2017/3D_slab_modes.py

```diff
@@ -0,0 +1,779 @@
+res = tuple(101 * np.array((16,9)))
+#res = tuple(51 * np.array((16,9)))
+#res = tuple(21 * np.array((16,9)))
+
+number_of_frames = 50
+
+# Frames per second of output video
+fps = 20
+
+#save_images = False
+save_images = True
+
+#make_video = False
+make_video = True
+
+#
+##
+###
+####
+#####
+######
+#######
+########
+#########
+
+# Variable definitions:
+# x = k*x
+# y = k*y
+# z = k*z
+# W = omega/k
+# K = k*x_0
+# t = omega*t
+
+if np.array([show_density, show_vel_front_pert, show_vel_top_pert]).any() == True and mode in alfven_mode_options:
+    raise NameError('Cannot show density or vel/disp pert for this mode')
+
+# Loop through selected modes
+for mode_ind in [0,1]:#range(8,14): # for all others. REMEMBER SBB pparameters
+#for mode_ind in [14,15]: #for fast body surf. REMEMBER SBS parameters
+#for mode_ind in [16, 17]:
+#for mode_ind in [13]: #for an individual mode
+#for mode_ind in range(2,14): 
+    if mode_ind not in range(len(mode_options)):
+        raise NameError('Mode not in mode_options')
+        
+    # (note that fast surface modes, i.e. 14 and 15, can only be 
+    # found with SBS parameters in slab_functions...)
+    mode = mode_options[mode_ind]    
+    
+    # Specify oscillation parameters
+    if mode in slow_surf_mode_options + alfven_mode_options:
+        K = 2.
+    elif mode in slow_body_mode_options:
+        K = 8.
+    elif mode in fast_body_1_mode_options:
+        K = 8.
+    elif mode in fast_body_2_mode_options:
+        K = 15.
+    elif mode in fast_body_3_mode_options:
+        K = 22.
+    elif mode in fast_surf_mode_options:
+        K = 8.
+    else:
+        raise NameError('Mode not found')
+
+# Specify density ratio R1 := rho_1 / rho_0
+#    R1 = 1.5 # Higher denisty on left than right
+#    R1 = 1.8
+#    R1 = 1.9 # Disp_diagram will only work for R1=1.5, 1.8, 2.0       
+    R1 = 2. # Symmetric slab
+    
+    # Reduce number of variables in dispersion relation
+    def disp_rel_asym_2var(W, K):
+        return sf.disp_rel_asym(W, K, R1)
+    
+    # find eigenfrequencies W (= omega/k) within the range Wrange for the given parameters.
+    Wrange1 = np.linspace(0., sf.cT, 11)
+    Wrange2 = np.linspace(sf.cT, sf.c0, 401)
+    Wrange3 = np.linspace(sf.c0, sf.c2, 11)
+    
+    Woptions_slow_surf = np.real(tool.point_find(disp_rel_asym_2var, np.array(K), Wrange1, args=None).transpose())
+    Woptions_slow_body = np.real(tool.point_find(disp_rel_asym_2var, np.array(K), Wrange2, args=None).transpose())
+    Woptions_fast = np.real(tool.point_find(disp_rel_asym_2var, np.array(K), Wrange3, args=None).transpose())
+    
+    # Remove W values that are very close to characteristic speeds - these are spurious solutions
+    tol = 1e-2
+    indices_to_rm = []
+    for i in range(len(Woptions_slow_surf)):
+        w = Woptions_slow_surf[i]
+        if min(abs(np.array([w, w - sf.c0, w - sf.c1(R1), w - sf.c2, w - sf.vA]))) < tol or w < 0 or w > sf.cT:
```

Comment:
Its basically making sure we only find the solutions to the disp relation that we want to find. In particular, we dont want any spurious solutions that arent actually eigenmodes, and we only want the first 3 kink and sausage body modes. I'll make it more clear.